agent/main.go

```diff
@@ -129,20 +115,48 @@ func main() {
 	err = wait.PollImmediate(time.Second, time.Minute, populateGroupList)
 	dieIf(err)
 
-	// create bcloud grpc api client and streams
+	// perform agent registration
 
-	var opts grpc.DialOption
+	id := os.Getenv("CLIENT_ID")
+	if *clientID != "" {
+		id = *clientID
+	}
+	if id == "" {
+		log.Fatal("missing client id! set -client-id flag or CLIENT_ID env var")
+	}
+	log.Debugf("using bcloud client id %s", id)
+
+	secret := os.Getenv("CLIENT_SECRET")
+	if *clientSecret != "" {
+		secret = *clientSecret
+	}
+	if secret == "" {
+		log.Fatal("missing bcloud client secret! set -client-secret flag or CLIENT_SECRET env var")
+	}
+
+	secure := !*insecure
+	bcloudApiClient := bcloudapi.New(id, secret, *apiAddr, secure)
+	agentRegistrator := registrator.New(bcloudApiClient, k8sAPI)
+
+	agentInfo, ts, err := agentRegistrator.EnsureRegistered(context.Background())
+	dieIf(err)
+	log.Infof("Obtained agent info: %+v", agentInfo)
+
+	// create bcloud grpc api client and streams
+	perRPCCreds := auth.NewTokenPerRPCCreds(ts, secure)
```

Comment:
it feels a bit brittle that we pass `secure` into `bcloudapi.New()`, which affects the `ts` returned by `agentRegistrator.EnsureRegistered`, but then we pass both `ts` and `secure` again into `auth.NewTokenPerRPCCreds()`.

instead, how about we only set secure in `bcloudapi.New()`, move `auth` into `bcloudapi`, and get `agentRegistrator` out of the business of retrieving a `TokenSource` from `bcloudapi` and passing it back to main? and then going a step further, we can replace `AgentTokenSource` with a new method that simply returns a `credentials.PerRPCCredentials`, so the caller never has to see a `oauth2.TokenSource`. lmkwyt:

```diff
diff --git a/agent/main.go b/agent/main.go
index fa4960d..0a58404 100644
--- a/agent/main.go
+++ b/agent/main.go
@@ -10,7 +10,6 @@ import (
 	"time"

 	"github.com/buoyantio/linkerd-buoyant/agent/pkg/api"
-	"github.com/buoyantio/linkerd-buoyant/agent/pkg/auth"
 	"github.com/buoyantio/linkerd-buoyant/agent/pkg/bcloudapi"
 	"github.com/buoyantio/linkerd-buoyant/agent/pkg/handler"
 	"github.com/buoyantio/linkerd-buoyant/agent/pkg/k8s"
@@ -138,12 +137,12 @@ func main() {
 	bcloudApiClient := bcloudapi.New(id, secret, *apiAddr, secure)
 	agentRegistrator := registrator.New(bcloudApiClient, k8sAPI)

-	agentInfo, ts, err := agentRegistrator.EnsureRegistered(context.Background())
+	agentInfo, err := agentRegistrator.EnsureRegistered(context.Background())
 	dieIf(err)
 	log.Infof("Obtained agent info: %+v", agentInfo)

 	// create bcloud grpc api client and streams
-	perRPCCreds := auth.NewTokenPerRPCCreds(ts, secure)
+	perRPCCreds := bcloudApiClient.Credentials(context.Background(), agentInfo.AgentID)
 	opts := []grpc.DialOption{grpc.WithPerRPCCredentials(perRPCCreds)}
 	if *insecure {
 		opts = append(opts, grpc.WithInsecure())
diff --git a/agent/pkg/bcloudapi/client.go b/agent/pkg/bcloudapi/client.go
index f1d317d..d6098be 100644
--- a/agent/pkg/bcloudapi/client.go
+++ b/agent/pkg/bcloudapi/client.go
@@ -10,6 +10,7 @@ import (
 	"github.com/buoyantio/linkerd-buoyant/agent/pkg/k8s"
 	"golang.org/x/oauth2"
 	"golang.org/x/oauth2/clientcredentials"
+	"google.golang.org/grpc/credentials"
 )

 const (
@@ -28,7 +29,7 @@ type AgentInfo struct {

 type Client interface {
 	RegisterAgent(ctx context.Context, agentName string) (*AgentInfo, error)
-	AgentTokenSource(ctx context.Context, agentID string) oauth2.TokenSource
+	Credentials(ctx context.Context, agentID string) credentials.PerRPCCredentials
 }

 // ApiClient is a client that talks to the Bcloud client API
@@ -38,6 +39,7 @@ type ApiClient struct {
 	tokenAuthConfig *clientcredentials.Config
 	addrScheme      string
 	apiAddr         string
+	secure          bool
 }

 // New creates a new ApiClient
@@ -60,6 +62,7 @@ func New(clientID, clientSecret, apiAddr string, secure bool) *ApiClient {
 		tokenAuthConfig: tokenAuthConfig,
 		addrScheme:      addrScheme,
 		apiAddr:         apiAddr,
+		secure:          secure,
 	}
 }

@@ -94,8 +97,8 @@ func (c *ApiClient) RegisterAgent(ctx context.Context, agentName string) (*Agent
 	return info, nil
 }

-// AgentTokenSource returns a token source for a particular agent
-func (c *ApiClient) AgentTokenSource(ctx context.Context, agentID string) oauth2.TokenSource {
+// Credentials returns a token source for a particular agent
+func (c *ApiClient) Credentials(ctx context.Context, agentID string) credentials.PerRPCCredentials {
 	authConfig := &clientcredentials.Config{
 		ClientID:     c.clientID,
 		ClientSecret: c.clientSecret,
@@ -103,5 +106,7 @@ func (c *ApiClient) AgentTokenSource(ctx context.Context, agentID string) oauth2
 		AuthStyle:    oauth2.AuthStyleInHeader,
 	}

-	return authConfig.TokenSource(ctx)
+	ts := authConfig.TokenSource(ctx)
+
+	return newTokenPerRPCCreds(ts, c.secure)
 }
diff --git a/agent/pkg/auth/token_auth.go b/agent/pkg/bcloudapi/creds.go
similarity index 64%
rename from agent/pkg/auth/token_auth.go
rename to agent/pkg/bcloudapi/creds.go
index df331be..9f30874 100644
--- a/agent/pkg/auth/token_auth.go
+++ b/agent/pkg/bcloudapi/creds.go
@@ -1,4 +1,4 @@
-package auth
+package bcloudapi

 import (
 	"context"
@@ -7,23 +7,23 @@ import (
 	"google.golang.org/grpc/credentials"
 )

-// TokenPerRPCCreds satisfies the credentials.PerRPCCredentials interface
+// tokenPerRPCCreds satisfies the credentials.PerRPCCredentials interface
 // attaching an up to date token to each RPC request in the GRPC pipeline.
 // When the token is expired a new token is requested by using the client
 // credentials flow.
-type TokenPerRPCCreds struct {
+type tokenPerRPCCreds struct {
 	secure bool
 	ts     oauth2.TokenSource
 }

-func NewTokenPerRPCCreds(ts oauth2.TokenSource, secure bool) credentials.PerRPCCredentials {
-	return &TokenPerRPCCreds{
+func newTokenPerRPCCreds(ts oauth2.TokenSource, secure bool) credentials.PerRPCCredentials {
+	return &tokenPerRPCCreds{
 		ts:     ts,
 		secure: secure,
 	}
 }

-func (t *TokenPerRPCCreds) GetRequestMetadata(ctx context.Context, uri ...string) (map[string]string, error) {
+func (t *tokenPerRPCCreds) GetRequestMetadata(ctx context.Context, uri ...string) (map[string]string, error) {
 	token, err := t.ts.Token()
 	if err != nil {
 		return nil, err
@@ -34,6 +34,6 @@ func (t *TokenPerRPCCreds) GetRequestMetadata(ctx context.Context, uri ...string
 	}, nil
 }

-func (t *TokenPerRPCCreds) RequireTransportSecurity() bool {
+func (t *tokenPerRPCCreds) RequireTransportSecurity() bool {
 	return t.secure
 }
diff --git a/agent/pkg/registrator/registrator.go b/agent/pkg/registrator/registrator.go
index a77c90d..4ed594d 100644
--- a/agent/pkg/registrator/registrator.go
+++ b/agent/pkg/registrator/registrator.go
@@ -8,7 +8,6 @@ import (
 	pkgk8s "github.com/buoyantio/linkerd-buoyant/agent/pkg/k8s"
 	"github.com/linkerd/linkerd2/pkg/k8s"
 	log "github.com/sirupsen/logrus"
-	"golang.org/x/oauth2"
 	kerrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
@@ -37,23 +36,23 @@ func New(bcloudApiClient bcloudapi.Client, k8sAPI *k8s.KubernetesAPI) *Registrat
 // EnsureRegistered performes agent registration if necessary. It inspects the
 // config map containing the agent metadata and decides whether this is a new agent
 // that needs to be registered.
-func (ar *Registrator) EnsureRegistered(ctx context.Context) (*bcloudapi.AgentInfo, oauth2.TokenSource, error) {
+func (ar *Registrator) EnsureRegistered(ctx context.Context) (*bcloudapi.AgentInfo, error) {
 	cm, err := ar.k8sAPI.
 		CoreV1().
 		ConfigMaps(pkgk8s.AgentNamespace).
 		Get(ctx, agentMetadataConfigMap, metav1.GetOptions{})
 	if err != nil {
 		if kerrors.IsNotFound(err) {
-			return nil, nil, fmt.Errorf("could not find %s config map", agentMetadataConfigMap)
+			return nil, fmt.Errorf("could not find %s config map", agentMetadataConfigMap)
 		}
-		return nil, nil, err
+		return nil, err
 	}

 	agentID, hasAgentID := cm.Data[pkgk8s.AgentIDKey]
 	agentName, hasAgentName := cm.Data[pkgk8s.AgentNameKey]

 	if !hasAgentName {
-		return nil, nil, fmt.Errorf("%s config map needs to have an %s key", agentMetadataConfigMap, pkgk8s.AgentNameKey)
+		return nil, fmt.Errorf("%s config map needs to have an %s key", agentMetadataConfigMap, pkgk8s.AgentNameKey)
 	}

 	if hasAgentID {
@@ -64,12 +63,12 @@ func (ar *Registrator) EnsureRegistered(ctx context.Context) (*bcloudapi.AgentIn
 			IsNewAgent: false,
 		}

-		return ai, ar.bcloudApiClient.AgentTokenSource(ctx, agentID), nil
+		return ai, nil
 	}

 	info, err := ar.bcloudApiClient.RegisterAgent(ctx, agentName)
 	if err != nil {
-		return nil, nil, fmt.Errorf("agent registration failed: %w", err)
+		return nil, fmt.Errorf("agent registration failed: %w", err)
 	}

 	cm.Data[pkgk8s.AgentIDKey] = info.AgentID
@@ -78,8 +77,8 @@ func (ar *Registrator) EnsureRegistered(ctx context.Context) (*bcloudapi.AgentIn
 		ConfigMaps(pkgk8s.AgentNamespace).
 		Update(ctx, cm, metav1.UpdateOptions{})
 	if err != nil {
-		return nil, nil, fmt.Errorf("failed to update %s configm map: %w", agentMetadataConfigMap, err)
+		return nil, fmt.Errorf("failed to update %s configm map: %w", agentMetadataConfigMap, err)
 	}

-	return info, ar.bcloudApiClient.AgentTokenSource(ctx, info.AgentID), nil
+	return info, nil
 }
diff --git a/agent/pkg/registrator/registrator_test.go b/agent/pkg/registrator/registrator_test.go
index adceafd..16d1100 100644
--- a/agent/pkg/registrator/registrator_test.go
+++ b/agent/pkg/registrator/registrator_test.go
@@ -125,7 +125,7 @@ func TestRegistrator(t *testing.T) {

 			apiClient := bcloudapi.New(clientID, clientSecret, apiAddr, false)
 			registrator := New(apiClient, k8sApi)
-			info, _, err := registrator.EnsureRegistered(context.Background())
+			info, err := registrator.EnsureRegistered(context.Background())
 			if err != nil {
 				if tc.expErr == nil {
 					t.Fatalf("Got unexpected error: %s", err)
```

Reply:
That is indeed much slicker, thanks